plugins/snapshots/windows/block_cimfs.go

```diff
@@ -0,0 +1,555 @@
+		"mount":         m,
+	}).Debugf("snapshot mounts")
+
+	return []mount.Mount{m}, nil
+}
+
+const (
+	layerCIMName    = "layer.cim"
+	layerBlockName  = "layer.vhd"
+	mergedCIMName   = "merged.cim"
+	mergedBlockName = "merged.vhd"
+)
+
+// prepareMergedCIM creates a new merged block CIM from the given snapshots. The order of
+// `snapshotIDs` is important. It is expected that the snapshot ID at the last index is
+// the base layer and the snapshot ID at the 0th index is the immediate parent layer.
+// Newly created merged CIM is stored in the same directory as that of the snapshot at 0th
+// index.  already exists in the 0th snapshot's directory, nothing is done.
+func (s *blockCIMSnapshotter) prepareMergedCIM(ctx context.Context, snapshotIDs []string) (rErr error) {
+	if len(snapshotIDs) < 2 {
+		return fmt.Errorf("merging CIM requires at least 2 snapshots")
+	}
+	log.G(ctx).WithFields(logrus.Fields{
+		"source snapshots": snapshotIDs,
+	}).Debugf("preparing merged CIM")
+
+	var (
+		// directory in which the merged CIM is stored
+		mergeDir          = s.getSnapshotDir(snapshotIDs[0])
+		mergeBlockCIMPath = filepath.Join(mergeDir, mergedBlockName)
+		snInfos           []snapshots.Info
+	)
+
+	// check if a merged CIM already exists
+	if _, err := os.Stat(mergeBlockCIMPath); err == nil {
+		log.G(ctx).Debugf("merged CIM already exists.")
+		return nil
+	}
+
+	err := s.ms.WithTransaction(ctx, false, func(tctx context.Context) error {
+		var terr error
+		snInfos, terr = s.snapshotInfoFromID(tctx, snapshotIDs)
+		if terr != nil {
+			return fmt.Errorf("failed to get info for snapshots: %w", terr)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	sourceCIMs := make([]*cimfs.BlockCIM, 0, len(snapshotIDs))
+	for i := 0; i < len(snapshotIDs); i++ {
+		sCIM, err := s.getSnapshotBlockCIM(ctx, snapshotIDs[i], snInfos[i])
+		if err != nil {
+			return err
+		}
+		sourceCIMs = append(sourceCIMs, sCIM)
+	}
+
+	mergedCIM := &cimfs.BlockCIM{
+		Type:      cimfs.BlockCIMTypeSingleFile,
+		BlockPath: mergeBlockCIMPath,
+		CimName:   "merged.cim",
+	}
+
+	// Build merge options based on snapshotter configuration
+	var mergeOpts []cimlayer.BlockCIMLayerImportOpt
+	if s.config.EnableLayerIntegrity {
+		mergeOpts = append(mergeOpts, cimlayer.WithLayerIntegrity())
+	}
+	if s.config.AppendVHDFooter {
+		mergeOpts = append(mergeOpts, cimlayer.WithVHDFooter())
+	}
+
+	err = cimlayer.MergeBlockCIMLayersWithOpts(ctx, sourceCIMs, mergedCIM, mergeOpts...)
+	if err != nil {
+		return fmt.Errorf("failed to merge CIMs: %w", err)
+	}
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"merged CIM": mergedCIM,
+	}).Debugf("merged CIM created")
+
+	return nil
+}
+
+func (s *blockCIMSnapshotter) isScratchSnapshot(snID string) (bool, error) {
+	scratchDiskPath := filepath.Join(s.getSnapshotDir(snID), "sandbox.vhdx")
+	if _, err := os.Stat(scratchDiskPath); err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
```

Comment:
Can we add a Cleanup() function like the other linux snapshotters have to take care of deleting any orphaned directories? I think we had decided on adding one for windows snapshotter for long term too.

Reply:
Sure we can do that.